Array chain rec

qlonik opened the pull request:
Addition of two implementations of chainRec (depth-first and breadth-first) for ReadonlyArray and Array.

In `ChainRec` instance by default the depth-first method is selected, but there is access to breadth-first via `BreadthFirstChainRec` instances.

As for the optimization of each of the two algorithms, breadth-first avoids initial array copy by having two loops that do the same thing. In depth-first I couldn't avoid copy of the initial array.

qlonik commented:
Friendly bump: is this useful?

gcanti commented:
@qlonik I'm sorry, somehow I missed this PR, I apologize.

Following the current naming convention (for example `SemigroupSum`, `SemigroupProduct`), I would rename: 

- `ChainRec` to `ChainRecDepthFirst`
- `BreadthFirstChainRec` to `ChainRecBreadthFirst`

(I'll retarget this PR to the `2.11` branch which is dedicated to the next release)

qlonik commented:
Done. I also rebased it for clean merge. 

In the `Array.ts` module, I'm reusing implementations from `ReadonlyArray.ts`. It used to be like that for other non-pipeable functions, which is why I implemented like this. However, it seems that now they are implemented independently between two modules. So, should I copy implementation from readonly array to regular array?

gcanti commented:
Thanks @qlonik
